File: app/src/main/java/com/mumanit/bontestapp/data/mappers/VenueDataMapperImpl.java
package com.mumanit.bontestapp.data.mappers;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.mumanit.bontestapp.data.api.rro.FoursquareFeaturedPhotos;
import com.mumanit.bontestapp.data.api.rro.FoursquareGroup;
import com.mumanit.bontestapp.data.api.rro.FoursquareJSON;
import com.mumanit.bontestapp.data.api.rro.FoursquareResults;
import com.mumanit.bontestapp.data.api.rro.FoursquareVenue;
import com.mumanit.bontestapp.data.api.rro.FoursquareVenuePhoto;
import com.mumanit.bontestapp.domain.model.VenueData;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by pmuciek on 7/8/17.
 */

public class VenueDataMapperImpl implements VenueDataMapper {

    @Override
    public VenueData map(@NonNull FoursquareVenue foursquareVenue) {

        String photoUrl = getPhotoUrl(foursquareVenue);
        String name = foursquareVenue.name;
        int checkinsCount = foursquareVenue.stats.checkinsCount;

        return new VenueData(checkinsCount, name, photoUrl);
    }

    @Override
    public List<VenueData> map(List<FoursquareVenue> foursquareVenueList) {

        List<VenueData> venueDataList = new ArrayList<>();

        for (FoursquareVenue foursquareVenue : foursquareVenueList) {
            venueDataList.add(map(foursquareVenue));
        }

        return venueDataList;
    }

    @Override
    public List<VenueData> map(FoursquareJSON foursquareJSON) {

        List<VenueData> result = new ArrayList<>();
        List<FoursquareGroup> foursquareGroups = foursquareJSON.response.groups;

        for (FoursquareGroup foursquareGroup : foursquareGroups) {
             for (FoursquareResults foursquareResults : foursquareGroup.items) {
                 result.add(map(foursquareResults.venue));
             }
        }

        return result;
    }

    private @Nullable String getPhotoUrl(@NonNull FoursquareVenue venue) {
        String venuePhotoUrl = null;

        FoursquareFeaturedPhotos photos = venue.featuredPhotos;

        if (null != photos && null != photos.items) {

            if (!photos.items.isEmpty()){
                FoursquareVenuePhoto photo = photos.items.get(0);
                venuePhotoUrl = photo.prefix + photo.suffix;
            }
        }

        return venuePhotoUrl;
    }
}

File: app/src/main/java/com/mumanit/foursquareclient/data/mappers/VenueDataMapperImpl.kt
package com.mumanit.foursquareclient.data.mappers

import com.mumanit.foursquareclient.data.api.rro.FoursquareJSON
import com.mumanit.foursquareclient.data.api.rro.FoursquareVenue
import com.mumanit.foursquareclient.domain.model.VenueDomainModel
import java.util.*

class VenueDataMapperImpl : VenueDataMapper {

    override fun map(foursquareVenue: FoursquareVenue): VenueDomainModel {
        val photoUrl = getPhotoUrl(foursquareVenue)
        val name = foursquareVenue.name
        val checkinsCount = foursquareVenue.stats.checkinsCount
        val id = foursquareVenue.id

        return VenueDomainModel(checkinsCount, name, photoUrl, id)
    }

    override fun map(foursquareVenueList: List<FoursquareVenue>): List<VenueDomainModel> {
        val venueDataList = ArrayList<VenueDomainModel>()

        for (foursquareVenue in foursquareVenueList) {
            venueDataList.add(map(foursquareVenue))
        }

        return venueDataList
    }

    override fun map(foursquareJSON: FoursquareJSON): List<VenueDomainModel> {
        val result = ArrayList<VenueDomainModel>()
        val foursquareGroups = foursquareJSON.response.groups

        for (foursquareGroup in foursquareGroups) {
            for (foursquareResults in foursquareGroup.items) {
                result.add(map(foursquareResults.venue))
            }
        }

        return result
    }

    private fun getPhotoUrl(venue: FoursquareVenue): String? {
        var venuePhotoUrl: String? = null
        val photos = venue.featuredPhotos

        if (photos?.items != null) {
            if (photos.items.isNotEmpty()) {
                val photo = photos.items[0]
                venuePhotoUrl = photo.prefix + "320x320" + photo.suffix
            }
        }

        return venuePhotoUrl
    }
}
